src/python/PyFunc.hpp

```diff
@@ -0,0 +1,95 @@
+/*
+ *~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * Copyright (c) 2019, Lawrence Livermore National Security, LLC.
+ *
+ * Produced at the Lawrence Livermore National Laboratory
+ *
+ * LLNL-CODE-746361
+ *
+ * All rights reserved. See COPYRIGHT for details.
+ *
+ * This file is part of the GEOSX Simulation Framework.
+ *
+ * GEOSX is a free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License (as published by the
+ * Free Software Foundation) version 2.1 dated February 1999.
+ *~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ */
+
+/**
+ * @file
+ */
+
+#pragma once
+
+// Source includes
+#include "pythonForwardDeclarations.hpp"
+#include "numpyHelpers.hpp"
+#include "PySortedArray.hpp"
+#include "../typeManipulation.hpp"
+#include "pythonHelpers.hpp"
+
+namespace LvArray
+{
+namespace python
+{
+
+namespace internal
+{
+
+bool err( void );
+
+void callPyFunc( PyObject * func, PyObject * * args, std::ptrdiff_t argc );
+
+} // namespace internal
+
+template< typename ... ARGS >
+class PythonFunction
+{
+public:
+
+  PythonFunction( PyObject * pyfunc ):
+    m_function( pyfunc )
+  { internal::xincref( pyfunc ); }
+
+  void operator()( ARGS ... args )
+  {
+    // if an error is already set, don't invoke the callback
```

Comment:
@jameshcorbett what is with this comment? It appears to me that the function is called if an error is set.

Reply:
I think we were arguing about what to do if the callback raises an error. If we just ignore it and try to invoke the callback again (since it may be invoked multiple times before returning control to the user's script) Python raises a `SystemError` which is not pretty and the actual exception is lost. Originally I suggested that once the callback raises, we just stop invoking it but continue as usual, and let the callback's exception (eventually) propagate to the user once the solver had finished. You didn't like that idea very much, for good reason. So we settled on raising a C++ exception (`throw PythonError();`) instead to return control to the user's script as soon as the callback raises. So the comment isn't entirely wrong in that you don't invoke the callback again once it has raised an error, but I think it should be reworded and moved down to the `throw` line.

Reply:
The usual thing to do would just be to propagate `nullptr` all the way back down to the Python C interface, but that wasn't an option for the physics solver.

Reply:
I remember this now.

That sounds like a good idea.